### PR TITLE
filestash: overhaul and fix package build

### PR DESCRIPTION
**Pull request by IndeedNotJames**
Tested the binary on x86_64-linux and aarch64-linux manually.
Everything seems to work, including image and thumbnails.
VM test builds as well.

Mostly copy-pasted from the commit messages:
- Filestash refers to go 1.20 in its `go.mod`, so we bump nixpkgs once, so we can use `buildGo120Module`
- Upstream uses go:embed now, so we need to use the frontend build as src
- `lib.getVersion` reports something useful now
- `git rev-parse HEAD` mock, which is called in `server/generator/constants.go`
- `excludedPackages` can be a list of strings
- Fix build date impurity
- Fix build for aarch64
- Remove `data/public` from the nixos module as the assets are now embedded

### CODE REVIEW

**Review by MatthewCroughan — CHANGES_REQUESTED** [review]
This looks really really good, a really good refactor, thank you so much!

File: pkgs/filestash/default.nix [hunk 1]
```diff
@@ -1,4 +1,18 @@
-{ dream2nix, stdenv, self, python, runCommand, pkgs, filestash-src, glib, vips, libraw, pkg-config, buildGoModule }:
+{ buildGo120Module
```
[the remark this answers is not in the document]
Comment (IndeedNotJames) [reply pkgs/filestash/default.nix:1]: 22.11's `buildGoModule` points to `buildGo119Module` and Filestash theoretically requires go 1.20 in https://github.com/mickael-kerjean/filestash/blob/26ee2006f4b0b5af5335e92238fd4a24c1a832c0/go.mod#L3

Filestash does however build just fine with `buildGo119Module` rn.
There just isn't any go 1.20 specific code in there yet.
This would more of an issue if Filestash wouldn't vendor the go dependencies afaik.

So yeah, dunno.
unstable and the upcoming 23.05 both use go 1.20 by default.
Not sure if switching back to unstable is worth it, though.
But that input isn't getting auto-updated anyway.

tl;dr: it doesn't matter right now, but it could in the future
 
Opinions?

---

lol I am getting http/500s trying to comment this :eyes:

Guess I have some time asking something else:
Do you think Filestash has a chance to get upstreamed into nixpkgs if Filestash starts providing a `package-lock.json`?

Update: https://www.githubstatus.com/incidents/52z0j6phhnjs has been resolved :)

[the remark this answers is not in the document]
Comment (IndeedNotJames) [reply pkgs/filestash/default.nix:1]: With that in mind, do you want me to change it back to `buildGoModule` or keep `buildGo120Module`? :)

[the remark this answers is not in the document]
Comment (MatthewCroughan) [reply pkgs/filestash/default.nix:1]: Yeah, in Nixpkgs for example they would want to use the latest buildGoModule, we should only pin the version if there is a legitimate observable compilation or runtime issue, same for all of the other libraries(inputs) involved

**Comment by IndeedNotJames** [comment]
ffs, the GitHub Action is still broken because I accidentally nuked `passthru.update = updateScript` (╯°□°）╯︵ ┻━┻

https://github.com/MatthewCroughan/filestash-nix/actions/runs/4591537178/jobs/8107771118

Can you fix this yourself, so we don't have to go through the whole PR workflow again?

**Comment by MatthewCroughan** [comment]
@IndeedNotJames fixed

**Comment by IndeedNotJames** [comment]
Wrong scope though.
Needs to be moved into `buildGoModule` to be accessible as `packages.x86_64-linux.filestash.passthru.update`

**Comment by MatthewCroughan** [comment]
Force pushed to master, that's embarassing.
